doc/btcli.1
.TH BTCLI "1" "2010\-07\-31" "BitTorrent Protocol Daemon 0.16" "User Commands"
.\" disable hyphenation
.nh
.\" adjust text to left margin only
.ad l
.\" -----------------------------------------------------------------
.\" MAIN CONTENT
.\" -----------------------------------------------------------------
.SH "NAME"
btcli \- command line interface to btpd
.SH "SYNOPSIS"
.B btcli
\fIOPERATION\fR [\fIOPERATION_OPTIONS\fR]
.SH "DESCRIPTION"
.PP
\fBbtcli\fR is a utility that interacts with a running btpd daemon.
It has several different modes of operation.
.SH "OPERATIONS"
.PP
One of the following operations must be specified when running btcli:
.TP
\fBadd\fR \- Add torrents to btpd.
.TP
\fBdel\fR \- Remove torrents from btpd.
.TP
\fBkill\fR \- Shut down btpd.
.TP
\fBlist\fR \- List torrents.
.TP
\fBrate\fR \- Set the global up and download rates in KB/s.
.TP
\fBstart\fR \- Activate torrents.
.TP
\fBstat\fR \- Display stats for active torrents.
.TP
\fBstop\fR \- Deactivate torrents.
.TP
\fB\-\-help\fR \fIOPERATION\fR Show help for the specified operation.
.SH "ADD OPTIONS"
.TP
\fB\-d\fR dir
Specifies the directory in which the files in the torrent will be downloaded;
It does not specify \fBbtpd\fR's "work" directory.
That is gathered from the environment variable \fI$HOME\fR.
.TP
\fB\-N, \-\-nostart\fR
Do not start the torrent immediately after adding it.
.TP
\fB\-T, \-\-topdir\fR
Append the torrent top directory (if any) to the content path.
.TP
\fB\-n\fR name
Specifies the name to be displayed for this torrent.
.TP
\fB\-l\fR label
Set the label to associate with torrent.
.SH "LIST OPTIONS"
.TP
\fB\-a\fR
List active torrents.
.TP
\fB\-i\fR
List inactive torrents.
.TP
\fB\-f\fR format
Specifies a custom format to use while displaying torrents.  The following is a
list of the characters that can be used:
.RS 8
.PP
\fB%n\fR \- torrent name
.br
\fB%#\fR \- torrent number
.br
\fB%h\fR \- torrent hash
.br
\fB%l\fR \- torrent label
.br
\fB%d\fR \- download directory
.br
\fB%t\fR \- state
.br
\fB%P\fR \- peer count
.PP
\fB%^\fR \- upload rate
.br
\fB%v\fR \- download rate
.PP
\fB%D\fR \- downloaded bytes
.br
\fB%U\fR \- uploaded bytes
.br
\fB%g\fR \- bytes got
.br
\fB%u\fR \- bytes uploaded
.br
\fB%s\fR \- total size, (formatted: prints K, M, G)
.br
\fB%S\fR \- total size, in bytes
.PP
\fB%A\fR \- available pieces
.br
\fB%T\fR \- total pieces
.br
\fB%H\fR \- have pieces
.PP
\fB%p\fR \- percent have (formatted)
.br
\fB%r\fR \- ratio
.PP
\fB%%\fR \- a percent symbol: '%'
.RE
.SH "STAT OPTIONS"
.TP
\fB\-i\fR
Display individual lines for each torrent.
.TP
\fB\-n\fR
Display the name of each torrent.  Implies '\-i'.
.SH "START OPTIONS"
.TP
\fB\-a\fR
Activate all non\-active torrents.
.SH "STOP OPTIONS"
.TP
\fB\-a\fR
Deactivate all torrents.
.SH "USAGE"
.PP
btpd must be started before btcli can be used.  See \fBbtpd\fR(1) for help with starting btpd.
.PP
To start sharing a torrent with btpd, the torrent needs to be added to btpd. This is done with \fBbtcli add\fR. Unless otherwise specified, btpd automatically starts to share the torrent and download any missing data. If the content directory you specify when adding a torrent does not exist, btpd will create it.
.PP
You can see which torrents have been added to btpd with \fBbtcli list\fR.  The list mode also displays a number for each added torrent. This number can be used to specify the target torrent for the btcli modes, so you don't have to keep the torrent file after adding it.
.PP
The up\- and download progress can be followed by using the \fBbtcli stat\fR mode. Both the list and stat modes use the following indicators to display the state of a torrent:
.RS 4
.TP 4
\fB+\fR
The torrent is starting. This may take time if btpd needs to test the content of this torrent or one started before it.
.TP 4
\fB\-\fR
The torrent is being stopped.
.TP 4
\fBI\fR
The torrent is inactive.
.TP 4
\fBS\fR
btpd is seeding the torrent.
.TP 4
\fBL\fR
btpd is leeching the torrent.
.RE
.PP
You can stop an active torrent with \fBbtcli stop\fR.  You can start an inactive torrent by using \fBbtcli start\fR.
.PP
.B Note:
Torrents can be specified either with its number or its file.
.PP
The \fBbtcli del\fR mode should only be used when you're totally finished with sharing a torrent. The mode will remove the torrent and its associated data from btpd. It is a bad idea to remove a not fully downloaded torrent and then add it again, since btpd has lost information on the not fully downloaded pieces and will need to download the data again.
.PP
To shut down btpd use \fBbtcli kill\fR.

.SH "EXAMPLES"
Display a list btpd's torrents and their number, size, status, etc.
.RS 4
.B $ btcli list
.RE
.PP
Same as above, but only for torrent 12 and my.little.torrent.
.br
.RS 4
.B $ btcli list 12 my.little.torrent
.RE
.PP
Same as above but only for active torrents.
.br
.RS 4
.B $ btcli list \-a
.RE
.PP
Add foo.torrent, with content dir foo.torrent.d, and start it.
.br
.RS 4
.B $ btcli add \-d foo.torrent.d foo.torrent
.RE
.PP
Same as above without starting it.
.br
.RS 4
.B $ btcli add \-\-no\-start \-d foo.torrent.d foo.torrent
.RE
.PP
Start bar.torrent and torrent number 7.
.br
.RS 4
.B $ btcli start bar.torrent 7
.RE
.PP
Stop torrent number 7.
.br
.RS 4
.B $ btcli stop 7
.RE
.PP
Stop all active torrents.
.br
.RS 4
.B $ btcli stop \-a
.RE
.PP
Remove bar.torrent and it's associated information from btpd.
.br
.RS 4
.B $ btcli del bar.torrent
.RE
.PP
Display a summary of up/download stats for the active torrents.
.br
.RS 4
.B $ btcli stat
.RE
.PP
Display the summary once every five seconds.
.br
.RS 4
.B $ btcli stat \-w 5
.RE
.PP
Same as above, but also display individual stats for each active torrent.
.br
.RS 4
.B $ btcli stat \-w 5 \-i
.RE
.PP
Set the global upload rate to 20KB/s and download rate to 1MB/s.
.br
.RS 4
.B $ btcli rate 20K 1M
.RE
.PP
Shut down btpd.
.br
.RS 4
.B $ btcli kill
.RE
.SH "BUGS"
Known bugs are listed at \fIhttp://github.com/btpd/btpd/issues\fR
.sp
Before submitting a bug report, please verify that you are running the latest version of btpd.
.SH "AUTHORS"
.sp
Current maintainers:
.sp
\- Marq Schneider <\fIqueueRAM@gmail.com\fR>
.sp
Past contributors:
.sp
\- Richard Nyberg <\fIbtpd@murmeldjur.se\fR> 
.SH "SEE ALSO"
.BR \fBbtpd\fR(1)
.BR \fBbtinfo\fR(1)



CHANGES
btpd 0.15:
----------

Bug fixes:
- The event loop could busy loop for up to a millisecond sometimes.
- On some platforms btpd couldn't initiate connections to other peers.
- The timer code was broken on MacOS X.

btpd 0.14:
----------

Misc:
- IPv6 support.
- Bans peers with bad data.
- Better tracker compatability and more reliable tracker communication.
- Show tracker counts in btcli stat.
- Compiles on Solaris 10.
- Btpd will now signal startup failures via the process exit value and
  print the failure reason to stdout.
- Better shutdown and torrent stop behaviour.
- Needs pthreads again.
- Removed libevent in favour of own event loop code.

Bug fixes:
- An integer array was misaligned, which could cause crashes on systems
  with alignment requirements. E.g., on XScale arm.
- An error in the torrent parsing code could cause crashes.
- The '--topdir' option to btcli add didn't work properly for single file
  torrents.
- The system name lookup scheme is now used. In btpd 0.13 only dns lookups
  were performed.

btpd 0.13:
----------

Misc:
- cURL is no longer required. btpd now uses its own http client.
- btpd doesn't use threads anymore, so pthreads are not required.
- Needed parts of libevent are now included in btpd.
- I/O errors for torrents no longer make btpd shut down. Instead the
  problematic torrent is stopped.
- Added the option '--ip' to specify the ip address btpd will report
  to trackers.
- Resume file format changed. Already added torrents will be retested once.
- The torrent info (ratio etc.) are now saved regularly.
- Reduced the amount of outgoing data kept in memory.

btpd 0.12:
----------

Misc:
- btpd now starts with the same set of active torrents as it had when it was
  last shut down. The new '--empty-start' option disables this feature.
- btpd now supports the key tracker extension.
- btpd now supports the multi tracker extension.
- btcli list and stat print headers for the presented data. Both commands
  displays more data than before.
- Torrent management has changed, with changes to the add and del commands
  and two new commands, start and stop, as a result. Operation on torrents
  can now be done by number as well as torrent file.
- A small delay has been added for contacting the same announce URL.
- btpd sends keepalive messages.
- The number of have messages sent has been reduced by not sending them to
  peers that already have the piece.
- The structure of the torrents directory in the btpd directory has changed.
  It is best to remove it before running btpd 0.12.

Bug fixes:
- Zero sized files in a torrent were not always created.

btpd 0.11:
----------

Bug fixes:
- A file descriptor was closed too early, which could cause btpd to exit
  with an error message on systems using epoll.

btpd 0.10:
----------

Misc:
- The protection mode of the ipc socket can now be specified via the new
  "--ipcprot" option for btpd.

Bug fixes:
- The pid file is now written correctly.
- btpd's "-d" option now works for relative paths too.
- Add a time out for handshake completion. Prevously btpd could slowly
  fill with unused incoming connections.

btpd 0.9:
---------

Misc:
- Small changes to 'btcli stat' output.

Bug fixes:
- Use correct value for '--prealloc' default.
- Stopped torrents now show correct rates (0kB/s).
- The test of content that has been modified between runs was broken.

btpd 0.8:
---------

Misc:
- btpd now requires pthreads.
- btpd keeps its logs and other information it needs in its own directory.
  $HOME/.btpd is the default location.
- Various changes in btpd options ant btcli use. See '--help'.
- The user chooses where to place content for each torrent.
- Better disk IO performance.
- Configureable disk preallocation to reduce file fragmentation.
- The unchoke algorithm is global instead of per torrent. The number
  of peers to upload to is configurable.
- Tracker errors are displayed in 'btcli stat' and most importantly don't
  make torrents unload on startup.
- The rates displayed by btcli are now computed in btpd and are averaged
  over the last 20 seconds.
- btpd now compares size and mtime of content to see if the resume file is
  valid.

Bug fixes:
- Double uninterest messages could be sent to some peers in endgame mode.
- A function was called prematurely which could trigger an assert in rare
  cases.

btpd 0.7:
---------

Bug fixes:
- Some communication stupidities was fixed, resulting in better download
  performance.

btpd 0.6:
---------

Misc:
- Set an upper limit on how much data a peer can request at once.
  This protects from running out of memory because of a greedy peer.
- End game is more efficient and uses much less memory.

btpd 0.5:
---------

Bug fixes:
- The bitfield could be sent too late. This could cause the receiving
  peer to close the connection.

btpd 0.4:
---------

Misc:
- The 0.1 rate limiter is back. It runs at a configurable hz now.
- Added "--bw-hz" option to btpd.
- Multiple have messages can be sent in place of a bitfield.
- If btpd can't write to a peer for 60s, it will be disconnected.

Bug fixes:
- A calculation error made btpd buggy with small torrents.
  This caused strange crashes and many reports on bad hashes in the log.
- The bitfield was sent too early. A peer could miss pieces btpd got
  while the peer was in handshake.
- Some situations in end game or with bad hashes could cause a crash.

btpd 0.3:
---------

Misc:
- New download strategy.
- Pipe ten requests instead of five.
- "-d" option added to btpd.
- "--with-warn" option added to configure. The default is now
  to compile with warnings off.

Bug fixes:
- It was possible for blocks to remain marked as busy when a peer
  choked or was lost, causing a piece download to wait until end game.
- The peer wanted level could get too high during end game.

btpd 0.2:
---------

Misc:
- A new way to do rate limiting.
- Compact mode is now used when talking to the tracker.

Bug fixes:
- A buffer was too small by four bytes, causing a buffer overflow

cli/btcli.c
#include <stdarg.h>

#include "btcli.h"
#include "utils.h"

const char *btpd_dir;
struct ipc *ipc;

void
diemsg(const char *fmt, ...)
{
    va_list ap;
    va_start(ap, fmt);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
    exit(1);
}

void
btpd_connect(void)
{
    if ((errno = ipc_open(btpd_dir, &ipc)) != 0)
        diemsg("cannot open connection to btpd in %s (%s).\n", btpd_dir,
            strerror(errno));
}

enum ipc_err
handle_ipc_res(enum ipc_err code, const char *cmd, const char *target)
{
    switch (code) {
    case IPC_OK:
        break;
    case IPC_COMMERR:
        diemsg("error in communication with btpd.\n");
    default:
        fprintf(stderr, "btcli %s '%s': %s.\n", cmd, target,
            ipc_strerror(code));
    }
    return code;
}

void
print_percent(long long part, long long whole)
{
    printf("%5.1f%% ", floor(1000.0 * part / whole) / 10);
}

void
print_rate(long long rate)
{
    if (rate >= 999.995 * (1 << 10))
        printf("%6.2fMB/s ", (double)rate / (1 << 20));
    else
        printf("%6.2fkB/s ", (double)rate / (1 << 10));
}

void
print_size(long long size)
{
    if (size >= 999.995 * (1 << 20))
        printf("%6.2fG ", (double)size / (1 << 30));
    else
        printf("%6.2fM ", (double)size / (1 << 20));
}
void
print_ratio(long long part, long long whole)
{
    printf("%7.2f ", (double)part / whole);
}

char
tstate_char(enum ipc_tstate ts)
{
    switch (ts) {
    case IPC_TSTATE_INACTIVE:
        return 'I';
    case IPC_TSTATE_START:
        return '+';
    case IPC_TSTATE_STOP:
        return '-';
    case IPC_TSTATE_LEECH:
        return 'L';
    case IPC_TSTATE_SEED:
        return 'S';
    }
    diemsg("unrecognized torrent state.\n");
}

int
torrent_spec(char *arg, struct ipc_torrent *tp)
{
    char *p;
    tp->u.num = strtoul(arg, &p, 10);
    if (*p == '\0') {
        tp->by_hash = 0;
        return 1;
    }
    if ((p = mi_load(arg, NULL)) == NULL) {
        fprintf(stderr, "btcli: bad torrent '%s' (%s).\n", arg,
            strerror(errno));
        return 0;
    }
    tp->by_hash = 1;
    mi_info_hash(p, tp->u.hash);
    free(p);
    return 1;
}

static struct {
    const char *name;
    void (*fun)(int, char **);
    void (*help)(void);
} cmd_table[] = {
    { "add", cmd_add, usage_add },
    { "del", cmd_del, usage_del },
    { "kill", cmd_kill, usage_kill },
    { "list", cmd_list, usage_list },
    { "rate", cmd_rate, usage_rate },
    { "start", cmd_start, usage_start },
    { "stop", cmd_stop, usage_stop },
    { "stat", cmd_stat, usage_stat }
};

static void
usage(void)
{
    printf(
        "btcli is the btpd command line interface.\n"
        "\n"
        "Usage: btcli [main options] command [command options]\n"
        "\n"
        "Main options:\n"
        "-d dir\n"
        "\tThe btpd directory.\n"
        "\n"
        "--help [command]\n"
        "\tShow this text or help for the specified command.\n"
        "\n"
        "Commands:\n"
        "add\t- Add torrents to btpd.\n"
        "del\t- Remove torrents from btpd.\n"
        "kill\t- Shut down btpd.\n"
        "list\t- List torrents.\n"
        "rate\t- Set up/download rate limits.\n"
        "start\t- Activate torrents.\n"
        "stat\t- Display stats for active torrents.\n"
        "stop\t- Deactivate torrents.\n"
        "\n"
        "Note:\n"
        "Torrents can be specified either with its number or its file.\n"
        "\n"
        );
    exit(1);
}

static struct option base_opts [] = {
    { "help", no_argument, NULL, 'H' },
    {NULL, 0, NULL, 0}
};

int
main(int argc, char **argv)
{
    int ch, help = 0;

    if (argc < 2)
        usage();

    while ((ch = getopt_long(argc, argv, "+d:", base_opts, NULL)) != -1) {
        switch (ch) {
        case 'd':
            btpd_dir = optarg;
            break;
        case 'H':
            help = 1;
            break;
        default:
            usage();
        }
    }
    argc -= optind;
    argv += optind;

    if (argc == 0)
        usage();

    if (btpd_dir == NULL)
        if ((btpd_dir = find_btpd_dir()) == NULL)
            diemsg("cannot find the btpd directory.\n");

    optind = 0;
    int found = 0;
    for (int i = 0; !found && i < ARRAY_COUNT(cmd_table); i++) {
        if (strcmp(argv[0], cmd_table[i].name) == 0) {
            found = 1;
            if (help)
                cmd_table[i].help();
            else
                cmd_table[i].fun(argc, argv);
        }
    }

    if (!found)
        usage();

    return 0;
}
